### DISCUSSION TITLE
Switch to a specific version of Python on macOS

### DISCUSSION THREAD

**Discussion by milosgajdos** [Homebrew/discussions#4857] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.1.17
ORIGIN: https://github.com/Homebrew/brew
HEAD: 35746e0a6ba6c3c5cfe56d99f79d9ec9f52ee15f
Last commit: 33 hours ago
Core tap JSON: 23 Oct 15:41 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_BROWSER: open
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: nvim
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.10_1/bin/ruby
CPU: octa-core 64-bit icelake
Clang: 14.0.0 build 1400
Git: 2.37.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 12.6.8-x86_64
CLT: 14.2.0.0.1.1668646533
Xcode: 14.2
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  aws-okta

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  aws-okta

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  docutils
  kubernetes-cli
```

### Description of issue

I am trying to switch python from `3.11` to `3.10` on my laptop

I thought running the `link` command would do the job:
```shell
$ brew link --overwrite python@3.10
Warning: Already linked: /usr/local/Cellar/python@3.10/3.10.13
To relink, run:
  brew unlink python@3.10 && brew link python@3.10
```

But that accomplishes nothing:
```shell
$ python3 -V
Python 3.11.6
```

**Comment by SMillerDev** [comment]
The easiest way to address a specific Homebrew version: ` python3.12 -V`
If you need a very specific version: save yourself some headache and use `pyenv`

**Comment by Bo98** [comment, marked as answer]
Yes, we support this but not via `brew link`. Add `$(brew --prefix python@3.10)/libexec/bin` to the front of your PATH and the default `python3` will change.

See `brew caveats python@3.10` for more info.

**Reply by milosgajdos** [reply]
Interesting. Thanks

Not sure about the `caveats` command though? I'm seeing this in `brew info python@3.10`
```
Python has been installed as
  /usr/local/bin/python3.10

Unversioned and major-versioned symlinks `python`, `python3`, `python-config`, `python3-config`, `pip`, `pip3`, etc. pointing to
`python3.10`, `python3.10-config`, `pip3.10` etc., respectively, have been installed into
  /usr/local/opt/python@3.10/libexec/bin

You can install Python packages with
  pip3.10 install <package>
They will install into the site-package directory
  /usr/local/lib/python3.10/site-packages

tkinter is no longer included with this formula, but it is available separately:
  brew install python-tk@3.10

If you do not need a specific version of Python, and always want Homebrew's `python3` in your PATH:
  brew install python3
```

**Reply by keeely** [reply]
So glad I found that little gem.  I was adding my own ~/bin path and adding symlinks in there just felt... ugly.

**Reply by soheil** [reply]
Why not provide a simple way to link to an older version without having to change my PATH everyone I run a bash cmd specially if you brew automatically updates python version making all previously install pip packages inaccessible?

**Reply by SMillerDev** [reply]
The simple way to link specific versions while keeping you packages is virtual environments. It's what the Python ecosystem is explicitly moving towards.

**Reply by soheil** [reply]
OK so it's even worse not only you have to reinstall all your packages you have to it in all your virtual envs too. $O(n^2)$ and it's not even the computer doing it.

**Reply by p-linnane** [reply]
Use `pyenv` or `pipx`. Homebrew's Python is mostly for other formulae that require it. You shouldn't be using it as a general purpose Python install.
